---
title: "class12"
author: "Kenny Dang (PID: A18544481)"
format: pdf
---

# Section 1. Proportion og G/G in a population 

Downloaded a CSV file from Ensemble < 

https://www.ensembl.org/Homo_sapiens/Variation/Sample?db=core;r=17:39800165-39990166;v=rs8067378;vdb=variation;vf=959672880;sample=MXL#373531_tablePanel

Here we read this CSV file

```{r}
mxl <- read.csv("HSdata.csv")
head(mxl)
```


```{r}
table(mxl$Genotype..forward.strand.)
```

```{r}
table(mxl$Genotype..forward.strand.)/nrow(mxl) * 100
```


Now let's look at a different population. I picked the GBR.

```{r}
gbr <- read.csv("GBRdata.csv")
```

Find proportion of G|G

```{r}
round(table(gbr$Genotype..forward.strand.)/nrow(gbr) * 100,2)
```
This variant that is associated with childhood asthma is more frequent in the GBR population than the MKL population. 

Let's now dig into this further. 

## Section 4: Population Scale Analysis 

One sample is obviously not enough to know what is happening in a population. You are
interested in assessing genetic differences on a population scale. 

So, you processed about
~230 samples and did the normalization on a genome level. Now, you want to find whether
there is any association of the 4 asthma-associated SNPs (rs8067378…) on ORMDL3
expression.

How many samples do we have?

```{r}
expr <- read.table("gen1.txt")

head(expr)
```

```{r}
nrow(expr)
```


```{r}
table(expr$geno)
```

```{r}
library(ggplot2)

```

Lets make a boxplot

```{r}
ggplot(expr) + aes(x=geno, y = exp, fill=geno) + geom_boxplot(notch=TRUE)
```


> Q13. Read this file into R and determine the sample size for each genotype and their corresponding median expression levels for each of these genotypes. 
Hint: The read.table(),summary() and boxplot() functions will likely be useful here. There is an example R script
online to be used ONLY if you are struggling in vein. Note that you can find the medium value from saving the output of the boxplot() function to an R object and examining this object. There is also the medium() and summary() function that you can use to check your understanding.

The total number of samples is 462. The sample size for the A/A genotype is 108 and its median expression level is 31.25. The sample size for the A/G genotype is 233 and its median expression level is 25.06. The sample size for the G/G genotype is 121 and its median expression level is 20.07. 

```{r}
expr <- read.table("gen1.txt")

head(expr)
```



```{r}
summary(expr)
```



```{r}
table(expr$geno)

```

```{r}
aggregate(exp ~ geno, data = expr, FUN = median)
```



```{r}
data.frame(
  genotype = names(tapply(expr$exp, expr$geno, median)),
  n        = as.vector(table(expr$geno)),
  median   = round(tapply(expr$exp, expr$geno, median, na.rm = TRUE), 2)
)
```





> Q14. Generate a boxplot with a box per genotype, what could you infer from the relative expression value between A/A and G/G displayed in this plot? Does the SNP effect the expression of ORMDL3? Hint: An example boxplot is provided overleaf – yours does not need to be as polished as this one.

The genotype A/A has the highest median expression level, then the genotype A/G has the second highest median expression level, and lastly the genotype G/G has the lowest median expression level. The median expression level decreases as the number of G alleles increase and so in other words, the G allele is associated with reduced gene expression. The boxplot indicates that there is a dose-dependent effect of the G allele. Yes, the SNP affects the expression of ORMDL3. ORMDL3 expression is highest in individuals with the A/A genotype, intermediate in individuals with the A/G genotype, and lowest in individuals with the G/G genotype.

```{r}
ggplot(expr, aes(x = geno, y = exp, fill = geno)) +
  geom_boxplot(outlier.shape = NA, alpha = 0.5) +
  geom_jitter(width = 0.2, color = "black", alpha = 0.3) +
  labs(
    title = "ORMDL3 Expression by Genotype",
    x = "Genotype",
    y = "Expression"
  )
```





