---
title: "Pathway Analysis from RNA-Seq Results"
author: "Kenny Dang (PID: A18544481)"
format: pdf
toc: true
---

## Section 1: Differential Expression Analysis

```{r}
library(DESeq2)
```



```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

colData = read.csv(metaFile, row.names=1)
head(colData)
```


```{r}
countData = read.csv(countFile, row.names=1)
head(countData)
```

> Q. Complete the code below to remove the troublesome first column from countData

```{r}
countData <- as.matrix(countData[,-1])
head(countData)
```


> Q. Complete the code below to filter countData to exclude genes (i.e. rows) where we have 0 read count across all samples (i.e. columns).

Tip: What will rowSums() of countData return and how could you use it in this context? 

```{r}
countData = countData[rowSums(countData) !=0,]
head(countData)
```

```{r}
dds = DESeqDataSetFromMatrix(countData = countData ,
                             colData= colData  ,
                             design =~ condition)

dds1 = DESeq(dds)
dds1
```


```{r}
res = results (dds1)
res
```

> Q. Call the summary() function on your results to get a sense of how many genes are up or down-regulated at the default 0.1 p-value cutoff. 

```{r}
summary (res)
```

```{r}
library (ggplot2)

ggplot(res) +
  aes(res$log2FoldChange,
  y= -log(res$padj)) +
  geom_point()
```

> Q. Improve this plot by completing the below code, which adds color, axis labels and cutoff lines: 

```{r}
mycols <- rep("gray", nrow(res)) 

# Color blue the genes with fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "blue"

# Color gray those with adjusted p-value more than 0.01
mycols[ res$padj > 0.01 ] <- "gray"

ggplot(res) +
  aes(x = log2FoldChange,
      y= -log(pvalue)) +
  geom_point(col = mycols) +
  xlab("Log2(FoldChange)") +
  ylab("-Log(P-value)") +
  geom_vline(xintercept = c(-2,2)) +
  geom_hline(yintercept = -log(0.01))
```


> Q. Use the mapIDs() function multiple times to add SYMBOL, ENTREZID and GENENAME annotation to our results by completing the code below.

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

res$symbol = mapIds(org.Hs.eg.db,
                    keys= row.names(res), 
                    keytype="ENSEMBL",
                    column= "SYMBOL",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")

head(res, 10)
```

> Q. Finally for this section let's reorder these results by adjusted p-value and save them to a CSV file in your current project directory.

```{r}
res = res[order(res$padj),]
write.csv(res, file  ="deseq_results.csv")
```



## Section 2. Pathway Analysis

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```


```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)
```



```{r}
head(keggres$less)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```


![](hsa04110.pathview.png)

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```


```{r}
keggrespathways <- rownames(keggres$greater)[1:5]

keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

![](hsa00140.pathview.png)









![](hsa04142.pathview.png)



![](hsa04630.pathview.png)


![](hsa04640.pathview.png)

![](hsa04330.pathview.png)




```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```



> Q. Can you do the same procedure as above to plot the pathview figures for the top 5 down-regulated pathways?

Yes.

```{r}
head(keggres$greater)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04640")

```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04640", kegg.native=T)

```





```{r}
keggrespathways1 <- rownames(keggres$less)[1:5]

keggresids1 = substr(keggrespathways1, start=1, stop=8)
keggresids1
```
![](hsa04110.pathview.png)

![](hsa03030.pathview.png)

![](hsa03013.pathview.png)

![](hsa03440.pathview.png)

![](hsa04114.pathview.png)

```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids1, species="hsa")
```



## Section 3 Gene Ontology (GO)

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets)

lapply(gobpres, head)
```
## Section 4. Reactome Analysis

```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```


> Q. What pathway has the most significant "Entities p-value"? Do the most significant pathways listed match your previous KEGG results? What factors could cause differences between the two methods?

The pathway that has the most significant "Entities p-value" is "Cell cycle, mitotic" as its value is 2E-5. Yes, the most significant pathways listed match my previous KEGG results as they both show cell-cycle related pathways as strongly enriched and they overlap. Some differences that could cause differences between the two methods are that they use different pathway annotations/gene groupings and they use different statistical models for testing enrichment. 



```{r}
sessionInfo()
```

